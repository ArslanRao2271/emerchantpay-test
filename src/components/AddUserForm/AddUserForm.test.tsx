import { render, screen } from "@testing-library/react";
import { AddUserForm } from ".";
import { MemoryRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import {
  emailFieldValidation,
  inputFields,
  nameFieldValidation,
  TestInputField,
} from "./utils";

const getWrapperComponent = (comp: JSX.Element) => {
  return render(comp, { wrapper: MemoryRouter });
};

describe("Add User Form test cases", () => {
  test("Should match previous snapshot", () => {
    const { asFragment } = getWrapperComponent(
      <AddUserForm getUserData={() => {}} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("Should not render any of the validation messages when component loads for first time", () => {
    getWrapperComponent(<AddUserForm getUserData={() => {}} />);

    expect(screen.queryByText(/Email is required/)).toBeNull();
    expect(screen.queryByText(/First name is required/)).toBeNull();
    expect(screen.queryByText(/Last name is required/)).toBeNull();
  });

  test("Should render all of the validation messages when someone click on submit without entering any information", async () => {
    getWrapperComponent(<AddUserForm getUserData={() => {}} />);

    expect(screen.queryByText(/Email is required/)).toBeNull();
    expect(screen.queryByText(/First name is required/)).toBeNull();
    expect(screen.queryByText(/Last name is required/)).toBeNull();

    const addUserButton = screen.getByRole("button", {
      name: /Add User/i,
    });

    userEvent.click(addUserButton);

    expect(await screen.findByText("Email is required")).toBeVisible();
    expect(await screen.findByText("First name is required")).toBeVisible();
    expect(await screen.findByText("Last name is required")).toBeVisible();
  });

  inputFields.forEach((item: TestInputField) =>
    test(`Should render error message "${item.textToFind}" if focus on "${item.inputField}" and then focus out`, async () => {
      getWrapperComponent(<AddUserForm getUserData={() => {}} />);

      const inputField = screen.getByPlaceholderText(item.inputField);

      expect(screen.queryByText(item.textToFind)).toBeNull();

      userEvent.click(inputField);
      userEvent.click(document.body);

      expect(await screen.findByText(item.textToFind)).toBeVisible();
    })
  );

  emailFieldValidation.forEach((item: string) =>
    test(`Should show validation message when invalid email "${item}" is provided `, async () => {
      getWrapperComponent(<AddUserForm getUserData={() => {}} />);

      const inputField = screen.getByPlaceholderText("Email address");

      expect(screen.queryByText("Email is required")).toBeNull();

      userEvent.click(inputField);
      userEvent.type(inputField, item);
      userEvent.click(document.body);

      expect(await screen.findByText("Enter a valid email")).toBeVisible();
    })
  );

  nameFieldValidation.forEach((item: string) =>
    test(`Should show validation message when invalid first name "${item}" is provided `, async () => {
      getWrapperComponent(<AddUserForm getUserData={() => {}} />);

      const inputField = screen.getByPlaceholderText("First Name");

      expect(screen.queryByText("First name is required")).toBeNull();

      userEvent.click(inputField);
      userEvent.type(inputField, item);
      userEvent.click(document.body);

      expect(
        await screen.findByText("Cannot contain special characters")
      ).toBeVisible();
    })
  );

  nameFieldValidation.forEach((item: string) =>
    test(`Should show validation message when invalid last name "${item}" is provided `, async () => {
      getWrapperComponent(<AddUserForm getUserData={() => {}} />);

      const inputField = screen.getByPlaceholderText("Last Name");

      expect(screen.queryByText("Last name is required")).toBeNull();

      userEvent.click(inputField);
      userEvent.type(inputField, item);
      userEvent.click(document.body);

      expect(
        await screen.findByText("Cannot contain special characters")
      ).toBeVisible();
    })
  );
});
