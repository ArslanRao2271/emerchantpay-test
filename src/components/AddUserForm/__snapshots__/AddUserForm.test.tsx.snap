// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`404 Page test cases Should match previous snapshot 1`] = `
<DocumentFragment>
  <form
    id="Add-user-form"
    novalidate=""
    style="display: flex; justify-content: center; align-items: center; flex-direction: column; gap: 1rem; width: 90%;"
  >
    <div
      class="MuiBox-root css-19szf4d"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-vht04n-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="on"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="email"
            name="email"
            placeholder="Email address"
            type="email"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-ihdtdm"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-vht04n-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="on"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="firstName"
            name="firstName"
            placeholder="First Name"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-ihdtdm"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-vht04n-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="on"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="lastName"
            name="lastName"
            placeholder="Last Name"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-ihdtdm"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-primary MuiButton-primaryPrimary MuiButton-sizeMedium MuiButton-primarySizeMedium css-n76phm-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="submit"
    >
      Add User
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </form>
</DocumentFragment>
`;

exports[`Add User Form Page test cases Should match previous snapshot 1`] = `
<DocumentFragment>
  <form
    id="Add-user-form"
    novalidate=""
    style="display: flex; justify-content: center; align-items: center; flex-direction: column; gap: 1rem; width: 90%;"
  >
    <div
      class="MuiBox-root css-19szf4d"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-vht04n-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="on"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="email"
            name="email"
            placeholder="Email address"
            type="email"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-ihdtdm"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-vht04n-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="on"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="firstName"
            name="firstName"
            placeholder="First Name"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-ihdtdm"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-vht04n-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="on"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="lastName"
            name="lastName"
            placeholder="Last Name"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-ihdtdm"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-primary MuiButton-primaryPrimary MuiButton-sizeMedium MuiButton-primarySizeMedium css-n76phm-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="submit"
    >
      Add User
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </form>
</DocumentFragment>
`;

exports[`Add User Form test cases Should match previous snapshot 1`] = `
<DocumentFragment>
  <form
    id="Add-user-form"
    novalidate=""
    style="display: flex; justify-content: center; align-items: center; flex-direction: column; gap: 1rem; width: 90%;"
  >
    <div
      class="MuiBox-root css-19szf4d"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-vht04n-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="on"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="email"
            name="email"
            placeholder="Email address"
            type="email"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-ihdtdm"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-vht04n-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="on"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="firstName"
            name="firstName"
            placeholder="First Name"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-ihdtdm"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-vht04n-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="on"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="lastName"
            name="lastName"
            placeholder="Last Name"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-ihdtdm"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-primary MuiButton-primaryPrimary MuiButton-sizeMedium MuiButton-primarySizeMedium css-n76phm-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="submit"
    >
      Add User
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </form>
</DocumentFragment>
`;
